amlb/uploads.py

```diff
@@ -0,0 +1,197 @@
+
+
+def _load_task_data(task_folder: pathlib.Path, fold: int = 0) -> Namespace:
+    """ Loads the metadata of the given fold of a task as a namespace. """
+    with open(task_folder / str(fold) / 'metadata.json', 'r') as fh:
+        metadata = json.load(fh)
+    metadata = Namespace.from_dict(metadata)
+    return metadata
+
+
+def _load_fold(task_folder: pathlib.Path, fold: int, task: OpenMLTask) -> pd.DataFrame:
+    """ Load the predictions and add openml repeat/fold/index information. """
+    prediction_file = task_folder / f"{fold}/predictions.csv"
+    predictions = pd.read_csv(prediction_file, sep=',', header=0)
+
+    train_indices, test_indices = task.get_train_test_split_indices(fold, repeat=0, sample=0)
+    predictions["index"] = test_indices
+    predictions["fold"] = fold
+    predictions["repeat"] = 0
+    return predictions
+
+
+def _load_predictions(task_folder: pathlib.Path) -> pd.DataFrame:
+    """ Loads predictions of all folds for a task with index information required for upload. """
+    metadata = _load_task_data(task_folder)
+    task = openml.tasks.get_task(metadata.openml_task_id)
+    results = [_load_fold(task_folder, fold, task) for fold in range(10)]
+    return pd.concat(results)
+
+
+def _list_completed_folds(task_folder: pathlib.Path) -> Set[str]:
+    completed_folds = set()
+    for fold_dir in [d for d in task_folder.iterdir() if d.is_dir()]:
+        if (task_folder / fold_dir / "predictions.csv").is_file():
+            completed_folds.add(fold_dir.name)
+    return completed_folds
+
+
+def _get_flow(metadata: Namespace, sync_with_server: bool = True) -> openml.flows.OpenMLFlow:
+    """ Creates or retrieves an OpenML flow for the given run metadata. """
+    if metadata.git_info.repo != "NA":
+        amlb_version = metadata.git_info.commit[:8] if metadata.git_info.tags == [] else metadata.git_info.tags[0]
+
+        *_, owner, _ = metadata.git_info.repo.rsplit('/')
+        if owner != 'openml':
+            amlb_version = f"{owner}#{amlb_version}"
+
+        flow_description = """
+            This flow is generated by {fork}the automl benchmark: {repo} 
+            Repository commit: {commit}
+            {framework_name} version: {framework_version}
+        """.format(
+            fork="a fork of " if owner != 'openml' else '',
+            repo=metadata.git_info.repo,
+            commit=metadata.git_info.commit,
+            framework_name=metadata.framework,
+            framework_version=metadata.framework_version,
+        )
+    else:
+        flow_description = """
+            This flow is generated by the automl benchmark: https://github.com/openml/automlbenchmark
+            Precise benchmark version information could not be determined.
+            {framework_name} version: {framework_version}
+        """.format(
+            framework_name=metadata.framework,
+            framework_version=metadata.framework_version
+        )
+        amlb_version = "NA"
```

Comment:
Should uploads be allowed in this case?

Reply:
I'm not sure if uploads should be allowed in such a case. By this we cannot make any statements about the results on OpenML as users could do any kinds of changes by accident.

Reply:
That's true, but users can technically already upload whatever they want. If the users make a mistake, it may still be preferred to upload the results with a clear indication that there are local changes (so as a unique flow). While sloppy, if a user had local changes running its experiments and has no time or resources to rerun the experiments, being able to upload and share them with an acknowledgement that it's not reproducible (or simply requires extra non-automated explanations) may be preferred in my opinion.